expression/builtin_compare.go

```diff
@@ -215,11 +255,556 @@ type compareFunctionClass struct {
+	if err != nil {
+		return sig, errors.Trace(err)
+	}
+	bf.tp.Flen = 1
+	intBf := baseIntBuiltinFunc{bf}
+	switch tp {
+	case tpInt:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTIntSig{intBf}
+		case opcode.LE:
+			sig = &builtinLEIntSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTIntSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQIntSig{intBf}
+		case opcode.GE:
+			sig = &builtinGEIntSig{intBf}
+		case opcode.NE:
+			sig = &builtinNEIntSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQIntSig{intBf}
+		}
+	case tpReal:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTRealSig{intBf}
+		case opcode.LE:
+			sig = &builtinLERealSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTRealSig{intBf}
+		case opcode.GE:
+			sig = &builtinGERealSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQRealSig{intBf}
+		case opcode.NE:
+			sig = &builtinNERealSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQRealSig{intBf}
+		}
+	case tpDecimal:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTDecimalSig{intBf}
+		case opcode.LE:
+			sig = &builtinLEDecimalSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTDecimalSig{intBf}
+		case opcode.GE:
+			sig = &builtinGEDecimalSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQDecimalSig{intBf}
+		case opcode.NE:
+			sig = &builtinNEDecimalSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQDecimalSig{intBf}
+		}
+	case tpString:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTStringSig{intBf}
+		case opcode.LE:
+			sig = &builtinLEStringSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTStringSig{intBf}
+		case opcode.GE:
+			sig = &builtinGEStringSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQStringSig{intBf}
+		case opcode.NE:
+			sig = &builtinNEStringSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQStringSig{intBf}
+		}
+	case tpDuration:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTDurationSig{intBf}
+		case opcode.LE:
+			sig = &builtinLEDurationSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTDurationSig{intBf}
+		case opcode.GE:
+			sig = &builtinGEDurationSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQDurationSig{intBf}
+		case opcode.NE:
+			sig = &builtinNEDurationSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQDurationSig{intBf}
+		}
+	case tpTime:
+		switch c.op {
+		case opcode.LT:
+			sig = &builtinLTTimeSig{intBf}
+		case opcode.LE:
+			sig = &builtinLETimeSig{intBf}
+		case opcode.GT:
+			sig = &builtinGTTimeSig{intBf}
+		case opcode.GE:
+			sig = &builtinGETimeSig{intBf}
+		case opcode.EQ:
+			sig = &builtinEQTimeSig{intBf}
+		case opcode.NE:
+			sig = &builtinNETimeSig{intBf}
+		case opcode.NullEQ:
+			sig = &builtinNullEQTimeSig{intBf}
+		}
+	}
+	return
+}
+
+type builtinLTIntSig struct {
+	baseIntBuiltinFunc
+}
+
+func (s *builtinLTIntSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
+	return compareInt(opcode.LT, s.args, row, s.ctx)
```

Comment:
`compareXXX` function will call `resOfCmp()` function.

for every `opcode.Op` we have write a corresponding `builtinXXSig` struct, and use that struct's `evalInt(...)` function to evaluate the final result. Here are 2 informations we can get when `builtinXXSig`'s `evalInt(...)` function being called:
- the compare type, i.e `opcode.Op`, can be: LT, LE, ...
- every parameter's return type: int, real, string ...

It's ok to use `compareXXX` to evaluate the two input parameters and compare them to get the result `rs`, which should be -1, 0, or 1. Here we use the parameter's return type information

But it's harmful to call `resOfCmp()`. Because in this function, we lost the information of compare type, and have to use a `switch ... case` to get the compare result, which should be true or false, for every possible compare type. This is inefficient

We can avoid this `switch ... case` by not call `resOfCmp()`. `compareXXX` should return the compare result `rs` and let caller to do what can be done in a concrete `case` branch of function `resOfCmp()`

`resOfCmp()` should be abandoned after refactoring,